File: sdk/typescript/src/review-commitment.ts
import { fromBech32, toBech32 } from "@cosmjs/encoding";
import { sha256 } from "@noble/hashes/sha2.js";
import type { ReviewAttestation } from "./generated/zerone/knowledge/v1/types";
import { MsgSubmitReveal } from "./generated/zerone/knowledge/v1/tx";

export interface ReviewCommitmentV2 {
  /** The immutable commitment_chain_id returned by the round, including after import. */
  chainId: string;
  roundId: string;
  verifier: string;
  vote: "accept" | "reject" | "malformed";
  confidence: bigint;
  salt: Uint8Array;
  attestation: ReviewAttestation;
}

const utf8 = new TextEncoder();
const strictUtf8 = new TextDecoder("utf-8", { fatal: true });
// Go unicode.IsSpace, used by strings.TrimSpace in the consensus validator.
const onlySpace = /^[\u0009-\u000d\u0020\u0085\u00a0\u1680\u2000-\u200a\u2028\u2029\u202f\u205f\u3000]*$/u;

function textBytes(name: string, value: string, max: number, required = false): Uint8Array {
  if (typeof value !== "string") throw new Error(`${name} must be a string`);
  const bytes = utf8.encode(value);
  if (bytes.length > max || strictUtf8.decode(bytes) !== value) {
    throw new Error(`${name} must be valid UTF-8 of at most ${max} bytes`);
  }
  if (required && onlySpace.test(value)) throw new Error(`${name} must be nonempty`);
  return bytes;
}

/**
 * Exact scheme-2 hash shared with x/knowledge/types.ComputeReviewCommitmentV2.
 * This binds a signer's assertion about a review; it proves neither independence
 * nor that the described work was performed. Never use it for scheme-0 rounds.
 */
export function computeReviewCommitmentV2(review: ReviewCommitmentV2): Uint8Array {
  const chain = textBytes("commitment chain", review.chainId, 256, true);
  const round = textBytes("commitment round", review.roundId, 256, true);
  const { prefix, data: address } = fromBech32(review.verifier);
  if (prefix !== "zrn" || address.length === 0 || address.length > 255 || toBech32(prefix, address) !== review.verifier) {
    throw new Error("reviewer must be a canonical Zerone account address");
  }
  if (!["accept", "reject", "malformed"].includes(review.vote)) throw new Error("invalid review vote");
  if (typeof review.confidence !== "bigint" || review.confidence < 0n || review.confidence > 1_000_000n) {
    throw new Error("review confidence must be between 0 and 1000000");
  }
  if (!(review.salt instanceof Uint8Array) || review.salt.length < 16 || review.salt.length > 64) {
    throw new Error("scheme-2 salt must contain 16 to 64 bytes");
  }
  const att = review.attestation;
  if (att === null || typeof att !== "object") throw new Error("review attestation is required");
  const known = new Set(["methodId", "reason", "evidenceIds", "scope"]);
  if (Object.keys(att).some((key) => !known.has(key))) throw new Error("unsupported review attestation fields");
  const method = textBytes("review method", att.methodId, 128);
  const reason = textBytes("review reason", att.reason, 4096, true);
  const scope = textBytes("review scope", att.scope, 1024);
  if (!Array.isArray(att.evidenceIds) || att.evidenceIds.length > 16) throw new Error("at most 16 evidence references are allowed");
  if (new Set(att.evidenceIds).size !== att.evidenceIds.length) throw new Error("duplicate evidence reference");
  const evidence = att.evidenceIds.map((value) => textBytes("evidence reference", value, 256, true));
  if (method.length + reason.length + scope.length + evidence.reduce((n, value) => n + value.length, 0) > 8192) {
    throw new Error("review attestation exceeds 8192 bytes");
  }
  const chunks: Uint8Array[] = [utf8.encode("ZRN.review.commit.v2\0")];
  const u32 = (value: number) => {
    const bytes = new Uint8Array(4);
    new DataView(bytes.buffer).setUint32(0, value, false);
    chunks.push(bytes);
  };
  const field = (bytes: Uint8Array) => { u32(bytes.length); chunks.push(bytes); };
  for (const value of [chain, round, address, utf8.encode(review.vote)]) field(value);
  const confidence = new Uint8Array(8);
  new DataView(confidence.buffer).setBigUint64(0, review.confidence, false);
  chunks.push(confidence);
  for (const value of [review.salt, method, reason, scope]) field(value);
  u32(evidence.length);
  for (const value of evidence) field(value);
  const preimage = new Uint8Array(chunks.reduce((n, value) => n + value.length, 0));
  let offset = 0;
  for (const value of chunks) { preimage.set(value, offset); offset += value.length; }
  return sha256(preimage);
}

/** Construct exactly the signed reveal fields whose scheme-2 hash was checked. */
export function makeReviewRevealV2(review: ReviewCommitmentV2): MsgSubmitReveal {
  computeReviewCommitmentV2(review);
  return MsgSubmitReveal.fromPartial({
    verifier: review.verifier, roundId: review.roundId, vote: review.vote,
    confidence: review.confidence, salt: Uint8Array.from(review.salt),
    attestation: { ...review.attestation, evidenceIds: [...review.attestation.evidenceIds] },
  });
}

File: x/knowledge/types/review_commitment.go
package types

import (
	"bytes"
	"crypto/sha256"
	"encoding/binary"
	"fmt"
	"strings"
	"unicode/utf8"

	sdk "github.com/cosmos/cosmos-sdk/types"
)

const (
	CommitmentSchemeLegacy    uint32 = 0
	CommitmentSchemeReviewV2  uint32 = 2
	MaxReviewReasonBytes             = 4096
	MaxReviewScopeBytes              = 1024
	MaxMethodIDBytes                 = 128
	MaxEvidenceReferences            = 16
	MaxEvidenceReferenceBytes        = 256
	MaxReviewAttestationBytes        = 8192
	MaxClaimReasoningBytes           = 8192
	MaxClaimReferences               = 64
	MaxCommitmentContextBytes        = 256
)

// ValidateRecordText bounds exact UTF-8 bytes without rewriting signed text.
func ValidateRecordText(name, value string, maxBytes int, required bool) error {
	if !utf8.ValidString(value) || len(value) > maxBytes {
		return fmt.Errorf("%s must be valid UTF-8 of at most %d bytes", name, maxBytes)
	}
	if required && strings.TrimSpace(value) == "" {
		return fmt.Errorf("%s must state a nonempty value", name)
	}
	return nil
}

func ValidateEvidenceReferences(refs []string) error {
	if len(refs) > MaxEvidenceReferences {
		return fmt.Errorf("at most %d evidence references are allowed", MaxEvidenceReferences)
	}
	seen := make(map[string]struct{}, len(refs))
	for _, ref := range refs {
		if err := ValidateRecordText("evidence reference", ref, MaxEvidenceReferenceBytes, true); err != nil {
			return err
		}
		if _, exists := seen[ref]; exists {
			return fmt.Errorf("duplicate evidence reference")
		}
		seen[ref] = struct{}{}
	}
	return nil
}

// ValidateReviewAttestation validates a statement of review, not its truth or quality.
func ValidateReviewAttestation(attestation *ReviewAttestation) error {
	if attestation == nil {
		return fmt.Errorf("scheme-2 review requires an attestation stating what was checked")
	}
	if len(attestation.ProtoReflect().GetUnknown()) != 0 {
		return fmt.Errorf("unsupported review attestation fields")
	}
	for _, field := range []struct {
		name, value string
		limit       int
		required    bool
	}{
		{"review reason", attestation.Reason, MaxReviewReasonBytes, true},
		{"review method", attestation.MethodId, MaxMethodIDBytes, false},
		{"review scope", attestation.Scope, MaxReviewScopeBytes, false},
	} {
		if err := ValidateRecordText(field.name, field.value, field.limit, field.required); err != nil {
			return err
		}
	}
	if err := ValidateEvidenceReferences(attestation.EvidenceIds); err != nil {
		return err
	}
	total := len(attestation.MethodId) + len(attestation.Reason) + len(attestation.Scope)
	for _, ref := range attestation.EvidenceIds {
		total += len(ref)
	}
	if total > MaxReviewAttestationBytes {
		return fmt.Errorf("review attestation exceeds %d bytes", MaxReviewAttestationBytes)
	}
	return nil
}

// CanonicalReviewAddress requires the exact SDK account-address spelling and
// returns its bytes. Ante authenticates this message signer independently.
func CanonicalReviewAddress(verifier string) ([]byte, error) {
	address, err := sdk.AccAddressFromBech32(verifier)
	if err != nil {
		return nil, fmt.Errorf("invalid reviewer address: %w", err)
	}
	if address.String() != verifier {
		return nil, fmt.Errorf("reviewer address must use canonical encoding")
	}
	return []byte(address), nil
}

// ComputeReviewCommitmentV2 hashes an unambiguous, versioned projection of all
// signed review fields. It uses the round's stored creation-chain identity;
// an imported in-flight round retains that identity, while its transaction
// is still authenticated on the importing chain by the SDK.
//
// Encoding: domain bytes "ZRN.review.commit.v2\x00", then uint32-big-endian
// length-prefixed chain, round, canonical reviewer bytes, vote; uint64-big-endian
// confidence; length-prefixed salt, method, reason, scope; uint32 evidence count
// and each length-prefixed evidence reference, preserving submitted order.
// No protobuf serialization, map iteration, delimiters, normalization or unknown
// fields participate. Legacy ComputeCommitmentHash is deliberately unchanged.
func ComputeReviewCommitmentV2(chainID, roundID, verifier, vote string, confidence uint64, salt []byte, attestation *ReviewAttestation) ([]byte, error) {
	if err := ValidateRecordText("commitment chain", chainID, MaxCommitmentContextBytes, true); err != nil {
		return nil, err
	}
	if err := ValidateRecordText("commitment round", roundID, MaxCommitmentContextBytes, true); err != nil {
		return nil, err
	}
	address, err := CanonicalReviewAddress(verifier)
	if err != nil {
		return nil, err
	}
	if vote != "accept" && vote != "reject" && vote != "malformed" {
		return nil, fmt.Errorf("invalid review vote")
	}
	if confidence > BPS {
		return nil, fmt.Errorf("review confidence must be between 0 and %d", BPS)
	}
	if len(salt) < 16 || len(salt) > 64 {
		return nil, fmt.Errorf("scheme-2 salt must contain 16 to 64 bytes")
	}
	if err := ValidateReviewAttestation(attestation); err != nil {
		return nil, err
	}
	h := sha256.New()
	_, _ = h.Write([]byte("ZRN.review.commit.v2\x00"))
	var word [8]byte
	put := func(value []byte) {
		binary.BigEndian.PutUint32(word[:4], uint32(len(value)))
		_, _ = h.Write(word[:4])
		_, _ = h.Write(value)
	}
	for _, value := range [][]byte{[]byte(chainID), []byte(roundID), address, []byte(vote)} {
		put(value)
	}
	binary.BigEndian.PutUint64(word[:], confidence)
	_, _ = h.Write(word[:])
	for _, value := range [][]byte{salt, []byte(attestation.MethodId), []byte(attestation.Reason), []byte(attestation.Scope)} {
		put(value)
	}
	binary.BigEndian.PutUint32(word[:4], uint32(len(attestation.EvidenceIds)))
	_, _ = h.Write(word[:4])
	for _, ref := range attestation.EvidenceIds {
		put([]byte(ref))
	}
	return h.Sum(nil), nil
}

// ValidateVerificationRoundRecord preserves the legacy scheme explicitly and
// refuses unknown scheme versions instead of selecting a fallback hash.
// Imported scheme-2 rounds keep their original chain context.
func ValidateVerificationRoundRecord(round *VerificationRound, recordIntegrityEnabled bool) error {
	if round == nil || round.Id == "" || round.ClaimId == "" {
		return fmt.Errorf("verification round requires id and claim_id")
	}
	if len(round.ProtoReflect().GetUnknown()) != 0 {
		return fmt.Errorf("unsupported verification round fields")
	}
	if round.Phase < VerificationPhase_VERIFICATION_PHASE_COMMIT || round.Phase > VerificationPhase_VERIFICATION_PHASE_EXPIRED {
		return fmt.Errorf("invalid verification round phase")
	}
	switch round.CommitmentScheme {
	case CommitmentSchemeLegacy:
		if round.CommitmentChainId != "" {
			return fmt.Errorf("legacy round cannot declare scheme-2 chain context")
		}
		for _, reveal := range round.Reveals {
			if reveal == nil || reveal.Confidence != 0 || reveal.Attestation != nil {
				return fmt.Errorf("legacy round cannot contain unbound review fields")
			}
		}
	case CommitmentSchemeReviewV2:
		if !recordIntegrityEnabled {
			return fmt.Errorf("scheme-2 round requires record-integrity activation")
		}
		if err := ValidateRecordText("commitment chain", round.CommitmentChainId, MaxCommitmentContextBytes, true); err != nil {
			return err
		}
		if err := ValidateRecordText("commitment round", round.Id, MaxCommitmentContextBytes, true); err != nil {
			return err
		}
		if len(round.Commits) > CommitSeatHardCap || len(round.Reveals) > CommitSeatHardCap || len(round.SelectedVerifiers) > CommitSeatHardCap {
			return fmt.Errorf("scheme-2 round exceeds verifier bound")
		}
		commits := make(map[string]*CommitEntry, len(round.Commits))
		for _, commit := range round.Commits {
			if commit == nil || len(commit.ProtoReflect().GetUnknown()) != 0 || len(commit.CommitHash) != sha256.Size {
				return fmt.Errorf("invalid scheme-2 commitment")
			}
			if _, err := CanonicalReviewAddress(commit.Verifier); err != nil {
				return err
			}
			if _, exists := commits[commit.Verifier]; exists {
				return fmt.Errorf("duplicate scheme-2 commitment")
			}
			commits[commit.Verifier] = commit
		}
		selected := make(map[string]bool, len(round.SelectedVerifiers))
		for _, verifier := range round.SelectedVerifiers {
			if selected[verifier] || commits[verifier] == nil {
				return fmt.Errorf("scheme-2 selected reviewers must match distinct commitments")
			}
			selected[verifier] = true
		}
		if len(selected) != len(commits) {
			return fmt.Errorf("scheme-2 selected reviewers omit a commitment")
		}
		seen := make(map[string]bool, len(round.Reveals))
		for _, reveal := range round.Reveals {
			if reveal == nil || len(reveal.ProtoReflect().GetUnknown()) != 0 {
				return fmt.Errorf("invalid scheme-2 reveal")
			}
			if seen[reveal.Verifier] {
				return fmt.Errorf("duplicate scheme-2 reveal")
			}
			seen[reveal.Verifier] = true
			commit := commits[reveal.Verifier]
			if commit == nil {
				return fmt.Errorf("scheme-2 reveal has no commitment")
			}
			hash, err := ComputeReviewCommitmentV2(round.CommitmentChainId, round.Id, reveal.Verifier, reveal.Vote, reveal.Confidence, reveal.Salt, reveal.Attestation)
			if err != nil {
				return err
			}
			if !bytes.Equal(commit.CommitHash, hash) {
				return fmt.Errorf("scheme-2 reveal does not match commitment")
			}
		}
	default:
		return fmt.Errorf("unsupported commitment scheme %d", round.CommitmentScheme)
	}
	if round.VerifierRewardSettlement != nil && !recordIntegrityEnabled {
		return fmt.Errorf("verifier settlement requires record-integrity activation")
	}
	return ValidateVerifierRewardSettlement(round)
}
